### PR TITLE
Use ResizeObserver to recalculate the font size for amp-fit-text fixes #29556. 

### LINKED ISSUE #29556: amp-fit-text: Content is unreadable when text style is adjusted
## WCAG Level
Level AA

## Priority
High

## Type
Adaptability

## Pages/Components affected
* [amp-fit-text](https://amp.dev/documentation/components/amp-social-share/)

## Description
An important design consideration is to make sure that the display of text can be adjusted, so that line height, spacing between paragraphs, and spacing between words and letters can be changed without the content becoming unreadable.

Currently, if a user adjusts their letter/word/line spacing, all variations of `<amp-fit-text>` - depending on the specific size defined for the component and the amount of content inside them - can end up cropping / clipping content.

Below are screenshots from the [examples page](https://amp.dev/documentation/examples/components/amp-fit-text/?format=websites), after the spacing metrics have been forcibly changed via this [text spacing bookmarklet](https://www.html5accessibility.com/tests/tsbookmarklet.html). Note that in all instances, the content of some of the boxes is clipped (either at the start, or end, or both, of the example sentence).

### Basic usage
Basic example with cropped content after spacing adjustment

### Aspect ratio
Aspect ratio example with cropped content after spacing adjustment

### min-font-size
min-font-size example with cropped content after spacing adjustment

### max-font-size
max-font-size example with cropped content after spacing adjustment

## User impact
Someone with a cognitive or reading disability like Dyslexia may find content easier to read when there is more space between letters, words, lines and paragraphs. If content becomes hidden or obscured when they adjust the way text is displayed it makes the problem worse not better.

In this case, depending on the size of the `<amp-fit-text>` element and the amount of content, users may end up being unable to read/perceive the information contained in the element.

## Required solution
Make sure that the element takes spacing metrics into account when doing its fit calculations. As users may change these spacing metrics after the page has loaded/the elements have rendered, this may require listening for changes and redoing the calculation/fit dynamically.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied:
1. Install a tool like the [ARC Toolkit](https://www.paciellogroup.com/toolkit/), which contains an option to change text spacing, or the [text spacing bookmarklet](https://www.html5accessibility.com/tests/tsbookmarklet.html).
2. Check that the text on the page responds to the changes.
3. Check that all content and functionality remains readable and available after the display of text has changed.

## Related standards
* [WCAG 2.1 Success Criterion 1.4.12 Text Spacing (Level AA)](https://www.w3.org/TR/WCAG21/#non-text-contrast)

## More information
* [W3C: Technique C35: Allowing for text spacing without wrapping](https://www.w3.org/WAI/WCAG21/Techniques/css/C35)
* [W3C: Technique C36: Allowing for text spacing override](https://www.w3.org/WAI/WCAG21/Techniques/css/C36)
* [Knowbility: Exploring WCAG 2.1: 1.4.12 Text Spacing](https://knowbility.org/blog/2018/WCAG21-1412TextSpacing/)
* [TPG: Short note on getting spaced out with SC 1.4.12 Text Spacing](https://developer.paciellogroup.com/blog/2018/05/short-note-on-getting-spaced-out-with-sc-1-4-12-text-spacing/)
* [TPG: ARC Toolkit](https://www.paciellogroup.com/toolkit/)

## Test data
Test date: July 2020

### PR DESCRIPTION
**Pull request by rebeccanthomas**
Add ResizeObserver to `amp-fit-text` which triggers `updateFontSize_` when the component children: `measurer_` or `content_` changes size. 
Add unit test to verify that `updateFontSize_` is appropriately called when the component size changes. 
This change fixes #29556.  

In addition to fixing this accessibility issue this change adds a new feature. The resizeObserver will now recalculate the font size if the content of the `amp-fit-text` component changes.

### CODE REVIEW

File: extensions/amp-fit-text/0.1/amp-fit-text.js [hunk 1]
```diff
@@ -104,6 +110,20 @@ class AmpFitText extends AMP.BaseElement {
         return this.textContent_ || this.contentWrapper_.textContent;
       },
     });
+
+    if (window.ResizeObserver) {
```
Comment (zhouyx) [inline extensions/amp-fit-text/0.1/amp-fit-text.js:114]: no need to register in `buildCallback` if you already have it in `layoutCallback`

File: extensions/amp-fit-text/0.1/amp-fit-text.js [hunk 2]
```diff
@@ -41,6 +41,12 @@ class AmpFitText extends AMP.BaseElement {
     /** @private {number} */
     this.maxFontSize_ = -1;
 
+    /** @private {!Array<unlistenDef>} */
+    this.unlisteners_ = [];
+
+    /** @private {ResizeObserver} */
+    this.observer_ = null;
```
Comment (zhouyx) [inline extensions/amp-fit-text/0.1/amp-fit-text.js:48]: There is no need to store the observer, also unlistener doesn't need to be an array. I recommend
```
this.resizeObserverUnlistener_ = null;
```
Then in `layoutCallback` and `unlayoutCallback` you check the `resizeObserverUnlistener_`

File: extensions/amp-fit-text/0.1/amp-fit-text.js [hunk 3]
```diff
@@ -118,11 +138,33 @@ class AmpFitText extends AMP.BaseElement {
 
   /** @override */
   layoutCallback() {
+    if (window.ResizeObserver) {
```
Comment (zhouyx) [inline extensions/amp-fit-text/0.1/amp-fit-text.js:141]: Use `this.win` instead.

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js [hunk 4]
```diff
@@ -71,6 +72,25 @@ describes.realWin(
...
+      ft.setAttribute('width', '50');
+      ft.setAttribute('height', '100');
+      ft.style.width = '50px';
+      ft.style.height = '100px';
+      const setupUpdateFontSizeSpy = env.sandbox.spy(
+        ft.implementation_,
+        'updateFontSize_'
+      );
+      await ft.implementation_.layoutCallback();
+      expect(setupUpdateFontSizeSpy).to.be.called;
```
Comment (zhouyx) [inline extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:92]: It's unclear whether the `updateFontSize_` is invoked by the resizeObserver or `layoutCallback` here. Better do
```
layoutCallback
// check called
change size
// check called
```

**Review by zhouyx — CHANGES_REQUESTED** [review]

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js (hunk 4 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:96]: I added the new tests, is this test more clear?

File: examples/amp-fit-text.amp.html [hunk 5]
```diff
@@ -0,0 +1,81 @@
+<!-- ## Introduction --><!--
+  [`amp-fit-text`](/content/amp-dev/documentation/components/reference/amp-fit-text-v0.1.md) enables you to manage the size and fit of text within a given area.
+--><!-- -->
+<!doctype html>
+<html ⚡>
+<head>
+<!-- ## Setup -->
+    <!--
+      Include the `amp-fit-text` component.
+    -->
```
Comment (caroqliu) [inline examples/amp-fit-text.amp.html:10]: These `<!-- ## -->` blocks are customarily elaborated for amp.dev code samples, but this file will not be ported to amp.dev. I'd recommend either adding this example in that repo or removing annotations from this file.

File: examples/amp-fit-text.amp.html [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    <!-- ## max-font-size -->
+    <!--
+      The `max-font-size` attribute dictates a maximum size for the text.  In this case, we've dictated a maximum of 10, ensuring it won't fill the size of its parent element.
+    -->
+    <div>
+      <div class="square-small"><amp-fit-text id="max small" width="200" height="200" layout="responsive" max-font-size="18">Lorem ipsum dolor sit amet, has nisl nihil convenire et, vim at aeque inermis reprehendunt.</amp-fit-text></div>
+      <div class="square-medium"><amp-fit-text id="max medium" width="200" height="200" layout="responsive" max-font-size="18">Lorem ipsum dolor sit amet, has nisl nihil convenire et, vim at aeque inermis reprehendunt.</amp-fit-text></div>
+      <div class="square-large"><amp-fit-text id="max large" width="200" height="200" layout="responsive" max-font-size="18">Lorem ipsum dolor sit amet, has nisl nihil convenire et, vim at aeque inermis reprehendunt.</amp-fit-text></div>
+    </div>
+</body></html>
```
Comment (caroqliu) [inline examples/amp-fit-text.amp.html:81]: nit: new line eof

File: extensions/amp-fit-text/0.1/amp-fit-text.js [hunk 7]
```diff
@@ -118,11 +121,31 @@ class AmpFitText extends AMP.BaseElement {
 
   /** @override */
   layoutCallback() {
+    if (this.win.ResizeObserver) {
```
Comment (caroqliu) [inline extensions/amp-fit-text/0.1/amp-fit-text.js:124]: please flip to save a level of indentation:
```js
if (!this.win.ResizeObserver) {
  return;
}
```

File: extensions/amp-fit-text/0.1/amp-fit-text.js [hunk 8]
```diff
@@ -41,6 +41,9 @@ class AmpFitText extends AMP.BaseElement {
     /** @private {number} */
     this.maxFontSize_ = -1;
 
+    /** @private {unlistenDef} */
+    this.resizeObserverUnlistener_ = null;
```
Comment (caroqliu) [inline extensions/amp-fit-text/0.1/amp-fit-text.js:45]: Should this be initialized to `() => {}` so that we can never accidentally call `null` in the `unlayoutCallback`?

**Review by caroqliu — CHANGES_REQUESTED** [review]

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js (hunk 4 above)
Comment (zhouyx) [reply extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:96]: `expect(setupUpdateFontSizeSpy).to.be.called;` once checks if the function has been called before. Use 
`calledOnce` and `calledTwice` to be more specific

File: extensions/amp-fit-text/0.1/amp-fit-text.js (hunk 8 above)
Comment (zhouyx) [reply extensions/amp-fit-text/0.1/amp-fit-text.js:45]: it is also used as a signal to tell if resizeObserver has been initialized. I recommend initialize it to `null` but add the check to `unlayoutCallback`

Comment (caroqliu) [reply extensions/amp-fit-text/0.1/amp-fit-text.js:45]: Works for me.

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js [hunk 9]
```diff
@@ -53,13 +53,14 @@ describes.realWin(
         .then(() => ft);
     }
 
-    it('renders', () => {
+    it('renders', async () => {
       const text = 'Lorem ipsum';
-      return getFitText(text).then((ft) => {
+      return getFitText(text).then(async (ft) => {
```
Comment (kristoferbaxter) [inline extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:58]: If this is moving to an async method, you can use `await` syntax instead of relying on the thenable directly.

File: extensions/amp-fit-text/0.1/amp-fit-text.js (hunk 8 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/amp-fit-text.js:45]: I set this to null to indicate that a resizeObserver has not been attached yet.

File: extensions/amp-fit-text/0.1/amp-fit-text.js (hunk 7 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/amp-fit-text.js:124]: Would it make more sense to just pull this part into the same if check as the following one? like 
```suggestion
    if (this.win.ResizeObserver && this.resizeObserverUnlistener_ === null) {
```

Comment (caroqliu) [reply extensions/amp-fit-text/0.1/amp-fit-text.js]: You still want to early return to save a level of indentation.
```js
if (!this.win.ResizeObserver || this.resizeObserverUnlistener_) {
  return;
}
```

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js [hunk 10]
```diff
@@ -53,14 +53,13 @@ describes.realWin(
...
       const text = 'Lorem ipsum';
-      return getFitText(text).then((ft) => {
-        const content = ft.querySelector('.i-amphtml-fit-text-content');
-        expect(content).to.not.equal(null);
-        expect(content.textContent).to.equal(text);
-        expect(ft.textContent).to.equal(text);
-      });
+      const ft = await getFitText(text);
+      const content = ft.querySelector('.i-amphtml-fit-text-content');
+      expect(content).to.not.equal(null);
```
Comment (caroqliu) [inline extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:60]: suggestion:
```suggestion
      expect(content).to.not.be.null;
```

File: extensions/amp-fit-text/0.1/amp-fit-text.js (hunk 7 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/amp-fit-text.js]: In this case we want to call the final return statement whether or not this if block is true, this specifically relates to deciding whether we should attach a new observer, would it be better to extract the logic which attaches the listener to a helper method? so that the layout callback would look like this?
`if (this.win.ResizeObserver && this.resizeObserverUnlistener_ === null) {
    attachNewResizeObserver();
}
return this.mutateElement(() => {
   this.updateFontSize_();
});`

Comment (caroqliu) [reply extensions/amp-fit-text/0.1/amp-fit-text.js]: Oh I see, apologies for missing that. In that case either way works :)

Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/amp-fit-text.js]: Sounds good, Thanks!

File: extensions/amp-fit-text/0.1/amp-fit-text.js (hunk 8 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/amp-fit-text.js:45]: Ah! I replied before I saw your comment, SGTM.

File: extensions/amp-fit-text/0.1/amp-fit-text.js [hunk 11]
```diff
@@ -118,11 +122,37 @@ class AmpFitText extends AMP.BaseElement {
 
   /** @override */
   layoutCallback() {
+    if (this.win.ResizeObserver && this.resizeObserverUnlistener_ === null) {
+      const observer = new this.win.ResizeObserver(() =>
+        throttle(
```
Comment (zhouyx) [inline extensions/amp-fit-text/0.1/amp-fit-text.js:127]: This is not how `throttle` method works. 
If you look at the implementation, `throttle` doesn't invoke the callback, it only wraps the given callback. 
I wonder why the test can pass.

**Review by zhouyx — CHANGES_REQUESTED** [review]

File: extensions/amp-fit-text/0.1/amp-fit-text.js (hunk 11 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/amp-fit-text.js:128]: Ah I see, I've modified this to wrap the callback entirely.

File: examples/amp-fit-text.amp.html [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      width:  200px;
+    }
+    .square-large {
+      height: 300px;
+      width:  300px;
+    }
+      </style>
+  </head>
+  <body>
+    <!--
```
Comment (caroqliu) [inline examples/amp-fit-text.amp.html:38]: Are these HTML comments necessary in the doc?

File: extensions/amp-fit-text/0.1/amp-fit-text.js [hunk 13]
```diff
@@ -41,6 +43,9 @@ class AmpFitText extends AMP.BaseElement {
     /** @private {number} */
     this.maxFontSize_ = -1;
 
+    /** @private {?unlistenDef} */
```
Comment (caroqliu) [inline extensions/amp-fit-text/0.1/amp-fit-text.js:46]: Should this be `function`? Where is `unlistenDef` defined?

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js [hunk 14]
```diff
@@ -53,14 +53,13 @@ describes.realWin(
...
-        const content = ft.querySelector('.i-amphtml-fit-text-content');
-        expect(content).to.not.equal(null);
-        expect(content.textContent).to.equal(text);
-        expect(ft.textContent).to.equal(text);
-      });
+      const ft = await getFitText(text);
+      const content = ft.querySelector('.i-amphtml-fit-text-content');
+      expect(content).to.not.be.null;
+      expect(ft.textContent).to.equal(text);
+      await ft.implementation_.unlayoutCallback();
```
Comment (caroqliu) [inline extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:62]: FMI: Why is this line added to test that the component renders?

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js [hunk 15]
```diff
@@ -71,6 +70,39 @@ describes.realWin(
...
+        'Lorem ipsum dolor sit amet, has nisl nihil convenire et, vim at aeque inermis reprehendunt.'
+      );
+      const updateFontSizeSpy = env.sandbox.spy(
+        ft.implementation_,
+        'updateFontSize_'
+      );
+      await new Promise((resolve) => {
+        setTimeout(() => {
+          resolve();
+        }, 90);
```
Comment (caroqliu) [inline extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:87]: Why `90`? Could be helpful to add an inline comment on why this value is meaningful.

File: examples/amp-fit-text.amp.html (hunk 12 above)
Comment (rebeccanthomas) [reply examples/amp-fit-text.amp.html:38]: I added them just for clarity and readability for anyone who would be looking at this file, but I can pull them out if they aren't necessary.

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js (hunk 15 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:87]: In this case `90` was arbitrarily chosen, i'll make it `100` and then add a comment to indicate why it was chosen.

Comment (caroqliu) [reply extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:87]: I wasn't sure if it was intentionally related to the throttle rate. Thanks for adding a comment :)

File: extensions/amp-fit-text/0.1/amp-fit-text.js (hunk 13 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/amp-fit-text.js:46]: UnlistenDef is defined in amp.extern.js (https://github.com/ampproject/amphtml/blob/master/build-system/externs/amp.extern.js#L414-L418) and refers to a function which specifically removes a listener, in this case the resizeObserver. I think its just a shade more precise than the `function` type declaration for this usage.

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js (hunk 15 above)
Comment (zhouyx) [reply extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:87]: The idea is to wait for the ResizeObserver to fire, but not until the next function call due to throttle. So any value less than 100 should be good.

Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:87]: It turns out 100ms exactly is a border case which can pass but its not clear whats happening so I am gonna re-set it to 90 and adjust my comments accordingly.

File: extensions/amp-fit-text/0.1/amp-fit-text.js (hunk 13 above)
Comment (caroqliu) [reply extensions/amp-fit-text/0.1/amp-fit-text.js:46]: Oh I see - then should this be capitalized?
```suggestion
    /** @private {?UnlistenDef} */
```

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js (hunk 15 above)
Comment (caroqliu) [reply extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:87]: Oh got it, thanks both for clarifying!

File: extensions/amp-fit-text/0.1/test/test-amp-fit-text.js (hunk 14 above)
Comment (rebeccanthomas) [reply extensions/amp-fit-text/0.1/test/test-amp-fit-text.js:62]: I initially added this because without this call the ResizeObserver unlistener, was not being correctly triggered, however after debugging this appears to have been related to a different issue so I have removed this line in this test.

**Review by zhouyx — APPROVED** [review]

**Review by caroqliu — APPROVED** [review]
